### ISSUE TITLE
Training Loss Different GPU vs. CPU

### ISSUE THREAD

**Issue by d-bohn** [mlverse/torch#481]
First off, thank you for the work that you all have done bringing `torch` to the R ecosystem. I've been wanting to try `torch` to supplement my `tensorflow`/`keras` deep learning with it for a while.

This issue may be due mostly to my ignorance of torch, but some Googling appears that others have experienced similar issues using `pytorch` and it is an unexpected behavior. Basically, I am trying to run a modified version of the example ("Guess the Correlation") and I get substantially different training losses when the only variable that has changed is `device = "cpu"` vs `device = "cuda"`. You can see the difference over 5 epochs below:

## GPU Training:

    Loss at epoch 1: training: 0.40000, validation: 0.38992

    Loss at epoch 2: training: 0.40521, validation: 0.40653
                                                                              
    Loss at epoch 3: training: 0.39901, validation: 0.39910

    Loss at epoch 4: training: 0.39461, validation: 0.39970
    
    Loss at epoch 5: training: 0.39562, validation: 0.38798

## CPU Training

    Loss at epoch 1: training: 0.13443, validation: 0.03077
    
    Loss at epoch 2: training: 0.01498, validation: 0.21862
    
    Loss at epoch 3: training: 0.01070, validation: 0.06492
    
    Loss at epoch 4: training: 0.00921, validation: 0.18520
    
    Loss at epoch 5: training: 0.00795, validation: 0.49545

I thought it may be due to an outdated NVIDIA driver, but I updated my driver and CUDA and the issue persisted (see info below). I am running this in a Docker container, but I have upgraded Docker to the latest version as well (though I don't see why that should affect this).

If this is a known phenomenon with Lantern, I apologize. However, if not, I would appreciate any ideas debugging the issue! Full script and GPU/CPU info below.

## Script
```{r eval = FALSE}
library(torch)
library(torchvision)
library(torchdatasets)

# device = "cpu"
device = "cuda"

# Get the datasets ----
train_indices <- 1:10000
val_indices <- 10001:15000
test_indices <- 15001:20000

add_channel_dim <- function(img) img$unsqueeze(1)
crop_axes <- function(img) transform_crop(img, top = 0, left = 21, height = 131, width = 130)

root <- file.path("data", "correlation")

train_ds <- guess_the_correlation_dataset(
  # where to unpack
  root = root,
  # additional preprocessing 
  transform = function(img) crop_axes(img) %>% add_channel_dim(),
  # don't take all data, but just the indices we pass in
  indexes = train_indices,
  download = TRUE
)

valid_ds <- guess_the_correlation_dataset(
  root = root,
  transform = function(img) crop_axes(img) %>% add_channel_dim(),
  indexes = val_indices,
  download = FALSE
)

test_ds <- guess_the_correlation_dataset(
  root = root,
  transform = function(img) crop_axes(img) %>% add_channel_dim(),
  indexes = test_indices,
  download = FALSE
)

# Work with batches ----
train_dl <- dataloader(train_ds, batch_size = 64, shuffle = TRUE)
batch <- dataloader_make_iter(train_dl) %>% dataloader_next()

par(mfrow = c(8,8), mar = rep(0, 4))

images <- as.array(batch$x$squeeze(2))

images %>%
  purrr::array_tree(1) %>%
  purrr::map(as.raster) %>%
  purrr::iwalk(~{plot(.x)})

valid_dl <- dataloader(valid_ds, batch_size = 64)
test_dl <- dataloader(test_ds, batch_size = 64)

# Create the model ----
torch_manual_seed(777)

net <- nn_module(
  
  "corr-cnn",
  
  initialize = function() {
    
    self$conv1 <- nn_conv2d(in_channels = 1, out_channels = 32, kernel_size = 3)
    self$conv2 <- nn_conv2d(in_channels = 32, out_channels = 64, kernel_size = 3)
    self$conv3 <- nn_conv2d(in_channels = 64, out_channels = 128, kernel_size = 3)
    
    self$bn1 <- nn_batch_norm2d(num_features = 32)
    self$bn2 <- nn_batch_norm2d(num_features = 64)
    self$bn3 <- nn_batch_norm2d(num_features = 128)
    self$bn4 <- nn_batch_norm1d(num_features = 128)
    
    self$fc1 <- nn_linear(in_features = 14 * 14 * 128, out_features = 128)
    self$fc2 <- nn_linear(in_features = 128, out_features = 1)
    
  },
  
  forward = function(x) {
    
    x %>% 
      self$conv1() %>%
      nnf_relu() %>%
      self$bn1() %>%
      nnf_avg_pool2d(2) %>%
      nnf_dropout(p = 0.2) %>%
      
      self$conv2() %>%
      nnf_relu() %>%
      self$bn2() %>%
      nnf_avg_pool2d(2) %>%
      nnf_dropout(p = 0.2) %>%
      
      self$conv3() %>%
      nnf_relu() %>%
      self$bn3() %>%
      nnf_avg_pool2d(2) %>%
      nnf_dropout(p = 0.2) %>%
      
      torch_flatten(start_dim = 2) %>%
      self$fc1() %>%
      nnf_relu() %>%
      self$bn4() %>%
      
      self$fc2()
  }
)

model <- net()

# Train the network ----
optimizer <- optim_adam(model$parameters)

train_batch <- function(b) {
  
  optimizer$zero_grad()
  
  # get predictions
  output <- model(b$x$to(device = device))
  
  # calculate loss
  loss <- nnf_mse_loss(output, b$y$unsqueeze(2)$to(device = device))
  
  # have gradients get calculated        
  loss$backward()
  
  # have gradients get applied
  optimizer$step()
  
  loss$item()
  
}

valid_batch <- function(b) {
  
  output <- model(b$x$to(device = device))
  loss <- nnf_mse_loss(output, b$y$unsqueeze(2)$to(device = device))
  loss$item()
  
}

model <- model$to(device = device)

num_epochs <- 5

for (epoch in 1:num_epochs) {
  
  cat("\nEpoch: ", epoch)
  
  # don't forget to do this
  model$train()
  
  train_losses <- c()
  
  pb <- progress::progress_bar$new(
    total = length(train_dl),
    format = " Training [:bar] :current/:total (:percent)"
    )
  for (b in enumerate(train_dl)) {
    pb$tick()
    loss <- train_batch(b)
    train_losses <- c(train_losses, loss)
    
  }
  
  # don't forget to do this either
  model$eval()
  
  valid_losses <- c()
  
  pb <- progress::progress_bar$new(
    total = length(valid_dl),
    format = " Validation [:bar] :current/:total (:percent)"
  )
  for (b in enumerate(valid_dl)) {
    pb$tick()
    loss <- valid_batch(b)
    valid_losses <- c(valid_losses, loss)
    
  }
  
  cat(sprintf("\nLoss at epoch %d: training: %1.5f, validation: %1.5f\n",
              epoch, mean(train_losses), mean(valid_losses)))
  
}
```

## Nvidia Driver and CUDA Info:

`$ nvidia-smi`

```{bash}
Mon Feb 22 20:06:26 2021
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 460.32.03    Driver Version: 460.32.03    CUDA Version: 11.2     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  TITAN RTX           Off  | 00000000:65:00.0  On |                  N/A |
| 41%   43C    P2    69W / 280W |  14295MiB / 24219MiB |      2%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
```

## CPU Info:
Architecture: x86_64
Model: Intel(R) Core(TM) i7-7800X CPU @ 3.50GHz

--------
EDIT

I have replicated these results using the official Docker image provided in the `torch` repository.

**Comment by dfalbel** [comment]
Sorry for the late reply. This is very unexpected.
I feel like them problem could be that `model <- model$to(device = device)` is called after `optimizer <- optim_adam(model$parameters)`, thus the optimizer is not pointing to the parameters on the gpu that are at this point used by the model.

**Comment by d-bohn** [comment]
Yup. Defining `model` after `optimizer` was the issue. Good to know that the order in which things are defined matters for `torch`. In retrospect, this make sense, but coming from a higher-level API where objects can be defined in any order up until they are all put together (training), I would not have figured that out. Thanks!

**Comment by dfalbel** [comment]
Yes! I agree this is the hardest part of torch. We should have a slightly higher level API in the future that takes care of these details! 

I am closing the issue! Feel free to open a new one if you have any additional questions.
